src/StorageFolding.cpp

```diff
@@ -789,22 +773,18 @@ class AttemptStorageFoldingOfFunction : public IRMutator {
                         to_release = max_required - max_required_next;  // This is the last time we use these entries
                     }
 
-                    // Logically we acquire the entire extent on
-                    // the first iteration:
-
-                    // to_acquire = select(loop_var > loop_min, to_acquire, extent);
-
-                    // However it's simpler to implement this by
-                    // just reducing the initial value on the
-                    // semaphore by the difference, as long as it
-                    // doesn't lift any inner names out of scope.
-
-                    Expr fudge = simplify(substitute(op->name, loop_min, extent - to_acquire));
-                    if (is_const(fudge) && can_prove(fudge <= sema.init)) {
-                        sema.init -= fudge;
-                    } else {
-                        to_acquire = select(loop_var > loop_min, likely(to_acquire), extent);
+                    if (provided.used.defined()) {
+                        to_acquire = select(provided.used, to_acquire, 0);
                     }
+                    if (required.used.defined()) {
```

Comment:
Don't you explicitly undefine required.used above?

Reply:
Yes, it's weird, added comment.

Reply:
Actually, I removed it. I was going to comment something to the effect of "this is correct, even though it's always a no-op". I think now it's actually wrong, for the same reason we clear required.used above, we still need to release the folded storage buffer even if we didn't use it.